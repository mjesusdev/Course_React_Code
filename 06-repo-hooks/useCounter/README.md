# useCounter Hook - React

Example:
```
    const { counter, increment, decrement, reset } = useCounter(10);
```

useCounter() // get value by default