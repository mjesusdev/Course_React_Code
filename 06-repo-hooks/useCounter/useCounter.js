import { useState } from "react";

export const useCounter = ( initialState = 10 ) => {   
    const [counter, setCounter] = useState(initialState);

    const reset = () => {
        setCounter( initialState );
    }

    const increment = () => {
        setCounter(counter + 1);
    }
    
    const decrement = () => {
        setCounter(counter - 1);
    }

    return {
        counter,
        increment,
        decrement,
        reset
    }
}

/*
export const useCounter = ( initialState ) => {   
    const [state, setState] = useState(initialState);

    const reset = () => {
        setState( initialState );
    }

    const increment = ( factor = 1 ) => {
        setState(state + factor);
    }
    
    const decrement = ( factor = 1 ) => {
        setState(state - factor);
    }

    return {
        state,
        increment,
        decrement,
        reset
    }
} */