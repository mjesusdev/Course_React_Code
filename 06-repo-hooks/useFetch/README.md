# useFetch Hook - React

Example:
```
    const url = 'endpoint API';
    const { data: null, loading: true, error: null } = useFetch(url);
```