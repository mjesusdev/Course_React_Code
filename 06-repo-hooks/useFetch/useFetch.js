import { useState, useEffect, useRef } from "react";

export const useFetch = ( url ) => {
    
    const isMounted = useRef(true);
    const [state, setState] = useState({ data: null, loading: true, error: null });

    useEffect( () => {
        return () => {
            isMounted.current = false;
        }
    }, []);

    useEffect( () => {

        setState({ data: null, loading: true, error: null });

        fetch( url ) 
            .then( resp => resp.json() )
            .then( data => {
                
                setTimeout( () => {
                    if ( isMounted.current ) {
                        setState({
                            loading: false,
                            error: null,
                            data
                        });
                    } else {
                        console.log('setState no se llamó');
                    }
                }, 2000);

            })
            .catch( () => {
                setState({
                    data: null,
                    loading: false,
                    error: 'No se puedo cargar la información'
                })
            })
    }, [url])

    return state;
}
