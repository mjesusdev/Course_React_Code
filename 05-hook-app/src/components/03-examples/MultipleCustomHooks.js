import React from 'react';
import { useFetch } from '../../hooks/useFetch';
import { useCounter } from '../../hooks/useCounter';

import '../02-useEffect/effects.css';

export const MultipleCustomHooks = () => {

    const { counter, increment } =  useCounter(1);
    const { loading, data } = useFetch( `https://www.breakingbadapi.com/api/quotes/${ counter }` );
    
    const { author, quote } = !!data && data[0];


    return (
        <div>
            <h1>BreakingBad Quotes</h1>
            <hr />


            {
                loading 
                ?
                    (
                        <div className="alert alert-info text-center">
                            Loading...
                        </div>
                    )
                :
                    (
                        <blockquote className="blockquote text-right">
                            <p className="mb-0"> { quote } </p>
                            <footer className="blockquote-footer"> { author } </footer>
                        </blockquote>
                    )
            }


            <button 
                className="btn btn-primary"
                onClick={ increment }
            >
                Siguiente quote
            </button>

        </div>
    )
}
