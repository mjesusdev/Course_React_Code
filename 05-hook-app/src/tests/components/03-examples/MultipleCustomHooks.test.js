import React from 'react';
import { shallow } from 'enzyme';
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';

jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');




describe('Pruebas en <MultipleCustomHooks />', () => {

    useCounter.mockReturnValue({
        counter: 10,
        increment: () => {}
    });

    
    test('debe de mostrarse correctamente', () => {

        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        });

        const wrapper = shallow( <MultipleCustomHooks /> );
        expect( wrapper ).toMatchSnapshot();
        
    })

    test('should debe de mostrar la informaciÃ³n', () => {
        
        useFetch.mockReturnValue({
            data: [{
                author: 'Fernando',
                quote: 'Hola Mundo'
            }],
            loading: false,
            error: null
        });

        const wrapper = shallow( <MultipleCustomHooks /> );

        expect( wrapper.find('.alert').exists() ).toBe( false );
        expect( wrapper.find('.mb-0').text().trim() ).toBe( 'Hola Mundo' );
        expect( wrapper.find('footer').text().trim() ).toBe( 'Fernando' );

    })
    
    

})


