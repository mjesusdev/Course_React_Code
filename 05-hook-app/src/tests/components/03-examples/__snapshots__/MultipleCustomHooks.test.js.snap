// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHooks /> debe de mostrarse correctamente 1`] = `
<div>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente quote
  </button>
</div>
`;
